File: 2590-maximum-star-sum-of-a-graph/2590-maximum-star-sum-of-a-graph.java


class Solution {
    public int maxStarSum(int[] vals, int[][] edges, int k) {
        // Initialize adjacency list for the graph
        List<List<Integer>> adj = new ArrayList<>();
        for (int i = 0; i < vals.length; i++) {
            adj.add(new ArrayList<>());
        }

        // Build adjacency list from edges
        for (int[] edge : edges) {
            adj.get(edge[0]).add(edge[1]);
            adj.get(edge[1]).add(edge[0]);
        }

        int maxsum = Integer.MIN_VALUE;

        // Iterate through each node
        for (int i = 0; i < vals.length; i++) {
            int sum = vals[i];  // Start the sum with the current node's value
            PriorityQueue<Integer> pq = new PriorityQueue<>(Collections.reverseOrder());

            // Add adjacent node values to the priority queue if they are positive
            for (int neighbor : adj.get(i)) {
                if (vals[neighbor] > 0) {
                    pq.offer(vals[neighbor]);
                }
            }

            // Take up to 'k' largest positive adjacent values
            int x = k;
            while (!pq.isEmpty() && x > 0) {
                sum += pq.poll();  // Add the largest positive value
                x--;
            }

            // Update the maximum sum
            maxsum = Math.max(maxsum, sum);
        }

        return maxsum;
    }
}

File: 2590-maximum-star-sum-of-a-graph/2590-maximum-star-sum-of-a-graph.cpp
class Solution {
public:
    int maxStarSum(vector<int>& vals, vector<vector<int>>& edges, int k) {
        vector<vector<int>> adj(vals.size());
       
        int maxsum = INT_MIN;
        for (int i = 0; i < edges.size(); i++) {
            adj[edges[i][0]].push_back(edges[i][1]);
            adj[edges[i][1]].push_back(edges[i][0]);
        }
        
        for (int i = 0; i < vals.size(); i++) {
            int sum = vals[i];
            priority_queue<int> pq;
           
            for (auto it : adj[i]) {
                pq.push(vals[it]);
            }
            int  x=k;
            while (!pq.empty()&&x > 0) {
                if(pq.top()<0)break;
                sum += pq.top();
                pq.pop();
                x--;
            }
         
            maxsum = max(maxsum, sum);
        }
        return maxsum;
    }
};